corgi/core/models.py

```diff
@@ -322,6 +324,17 @@ def save_product_taxonomy(self):
 
         return None
 
+    def disassociate_with_product(self, product_model_name: str, product_pk: str):
+        """Remove the product references from all components associated with this build."""
+        product_model = apps.get_model("core", product_model_name)
+        product = product_model.objects.get(pk=product_pk)
+
+        for component in self.components.iterator():
+            component.disassociate_with_product(product)
+            for cnode in component.cnodes.iterator():
+                for descendant in cnode.get_descendants().iterator():
+                    descendant.obj.disassociate_with_product(product)
```

Comment:
Is there a reason we iterate over cnodes? This requires two lookups, one to find the cnode's linked object's content_type, then another to find the cnode's linked object itself (which is always a Component). We can just iterate over component.provides and component.upstreams instead, right?
```suggestion
        for root_component in self.components.iterator():
            root_component.disassociate_with_product(product)
            for provided_component in root_component.provides.iterator():
                    provided_component.disassociate_with_product(product)
            for upstream_component in root_component.upstreams.iterator():
                    upstream_component.disassociate_with_product(product)
```

Reply:
I copied this code from save_product_taxonomy models.py 314-323. I hesitate to deviate from what's used there as this new function is designed to undo the changes made in that function.